lib/Mojolicious/Plugin/DefaultHelpers.pm

```diff
@@ -49,20 +49,34 @@ sub _asset {
   $c->rendered;
 }
 
-sub _block { ref $_[0] eq 'CODE' ? $_[0]() : $_[0] }
+sub _block {
+  my $content = shift;
+  ref $content eq 'CODE' ? $content->(@_) : $content;
+}
 
 sub _content {
-  my ($append, $replace, $c, $name, $content) = @_;
+  my ($append, $replace, $c, $name, @args) = @_;
   $name ||= 'content';
 
+  my $content = pop @args
+    if @args && ref $args[-1] eq 'CODE' || @args == 1 && ref $args[-1] eq '';
+
   my $hash = $c->stash->{'mojo.content'} ||= {};
   if (defined $content) {
-    if ($append) { $hash->{$name} .= _block($content) }
-    if ($replace) { $hash->{$name} = _block($content) }
-    else          { $hash->{$name} //= _block($content) }
+    if ($append) { push @{$hash->{$name}}, $content }
+    if ($replace) { $hash->{$name} = [_block($content, @args)] }
+    else          { $hash->{$name} //= [$content] }
+  }
+
+  return '' if !defined $hash->{$name} || !defined wantarray;
```

Comment:
This line is just speed improvement. It can be safely removed in prejudice of speed.